File: jane_street/electronic_trading_challenge/cpp/src/exchange.cpp
#include "etc/exchange.h"
#include <iostream>
#include <stdexcept>

#ifdef _WIN32
#include <winsock2.h>
#include <ws2tcpip.h>
#pragma comment(lib, "ws2_32.lib")
#else
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <unistd.h>
#endif

namespace etc {

void ExchangeConnection::send_hello(const std::string& team_name) {
    nlohmann::json msg = {{"type", "hello"}, {"team", team_name}};
    write(msg);
}

void ExchangeConnection::place_order(int order_id, const std::string& symbol, const std::string& side, int price, int size) {
    nlohmann::json msg = {
        {"type", "add"},
        {"order_id", order_id},
        {"symbol", symbol},
        {"dir", side},
        {"price", price},
        {"size", size}
    };
    write(msg);
}

void ExchangeConnection::cancel_order(int order_id) {
    nlohmann::json msg = {{"type", "cancel"}, {"order_id", order_id}};
    write(msg);
}

void ExchangeConnection::convert(int order_id, const std::string& symbol, const std::string& side, int size) {
    nlohmann::json msg = {
        {"type", "convert"},
        {"order_id", order_id},
        {"symbol", symbol},
        {"dir", side},
        {"size", size}
    };
    write(msg);
}

TcpExchangeConnection::~TcpExchangeConnection() {
    close();
}

void TcpExchangeConnection::connect(const std::string& host, int port) {
#ifdef _WIN32
    WSADATA wsaData;
    WSAStartup(MAKEWORD(2, 2), &wsaData);
#endif

    struct addrinfo hints{}, *res;
    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_STREAM;
    if (getaddrinfo(host.c_str(), std::to_string(port).c_str(), &hints, &res) != 0) {
        throw std::runtime_error("getaddrinfo failed");
    }

    sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
    if (::connect(sock, res->ai_addr, res->ai_addrlen) < 0) {
        freeaddrinfo(res);
        throw std::runtime_error("connect failed");
    }
    freeaddrinfo(res);
}

void TcpExchangeConnection::close() {
    if (sock != -1) {
#ifdef _WIN32
        closesocket(sock);
        WSACleanup();
#else
        ::close(sock);
#endif
        sock = -1;
    }
}

std::optional<nlohmann::json> TcpExchangeConnection::read() {
    char buf[1024];
    while (buffer.find('\n') == std::string::npos) {
        int n = recv(sock, buf, sizeof(buf), 0);
        if (n <= 0) return std::nullopt;
        buffer.append(buf, n);
    }
    size_t pos = buffer.find('\n');
    std::string line = buffer.substr(0, pos);
    buffer.erase(0, pos + 1);
    return nlohmann::json::parse(line);
}

void TcpExchangeConnection::write(const nlohmann::json& message) {
    std::string s = message.dump() + "\n";
    send(sock, s.c_str(), s.size(), 0);
}

} // namespace etc

File: jane_street/electronic_trading_challenge/rust/src/exchange.rs
use anyhow::Result;
use serde_json::{json, Value};
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::TcpStream;

pub struct ExchangeConnection {
    reader: BufReader<tokio::net::tcp::OwnedReadHalf>,
    writer: tokio::net::tcp::OwnedWriteHalf,
}

impl ExchangeConnection {
    pub async fn connect(host: &str, port: u16) -> Result<Self> {
        let stream = TcpStream::connect(format!("{}:{}", host, port)).await?;
        let (read, write) = stream.into_split();
        Ok(Self {
            reader: BufReader::new(read),
            writer: write,
        })
    }

    pub async fn read(&mut self) -> Result<Option<Value>> {
        let mut line = String::new();
        let n = self.reader.read_line(&mut line).await?;
        if n == 0 {
            return Ok(None);
        }
        Ok(Some(serde_json::from_str(&line)?))
    }

    pub async fn write(&mut self, message: Value) -> Result<()> {
        let mut data = serde_json::to_string(&message)?;
        data.push('
');
        self.writer.write_all(data.as_bytes()).await?;
        Ok(())
    }

    pub async fn send_hello(&mut self, team: &str) -> Result<()> {
        self.write(json!({"type": "hello", "team": team})).await
    }

    pub async fn place_order(&mut self, id: i32, sym: &str, dir: &str, price: i32, size: i32) -> Result<()> {
        self.write(json!({
            "type": "add",
            "order_id": id,
            "symbol": sym,
            "dir": dir,
            "price": price,
            "size": size
        })).await
    }

    pub async fn convert(&mut self, id: i32, sym: &str, dir: &str, size: i32) -> Result<()> {
        self.write(json!({
            "type": "convert",
            "order_id": id,
            "symbol": sym,
            "dir": dir,
            "size": size
        })).await
    }
}
